File: 프로그래머스/lv2_기능개발.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

vector<int> solution(vector<int> progresses, vector<int> speeds) {
    vector<int> answer;
    int num=0;
    int size = progresses.size();

    while( !progresses.empty() ){
        for(int i=progresses.size()-1; i>=0; i--){
            progresses[i] += speeds[i];

            // 맨앞 원소가 100보다 커지면, 앞에서부터 차례로 100이상인 수를 체크함
            if( progresses.front() >= 100 ){
                num=0;
                while( !progresses.empty() && progresses.front() >= 100 ){
                    num++;
                    progresses.erase(progresses.begin());
                    speeds.erase(speeds.begin());
                }
                answer.push_back(num);
            }
        }
    }
 
    return answer;
}

int main(void){
    vector<int> progresses = {95, 90, 99, 99, 80, 99};
    vector<int> speeds = {1, 1, 1, 1, 1, 1};   
    vector<int> ret;

    ret = solution(progresses, speeds);

    for(int i=0; i<ret.size(); i++){
        cout << ret[i] << " " ;
    }
    cout << endl;

    return 0;
}

File: 프로그래머스/lv2_기능개발.py
from collections import deque

def solution(progresses, speeds):
    answer = []

    deq_progresses = deque(progresses)
    deq_speeds = deque(speeds)

    while(len(deq_progresses) != 0):
        for i in range(0, len(deq_progresses)):
            deq_progresses[i] += deq_speeds[i]

        cnt = 0
        while( len(deq_progresses) != 0 and deq_progresses[0] >= 100 ):
            cnt += 1
            deq_progresses.popleft()
            deq_speeds.popleft()
        if( cnt != 0 ):
            answer.append(cnt)
        
    return answer

if __name__ == "__main__":
    progresses = [93, 30, 55]
    speeds = [1, 30, 5]

    ret = solution(progresses, speeds)
    print(ret)